### PR TITLE
Add GRPc Query for Upgrade

### PR DESCRIPTION
**Pull request by chatton**
## Description

~for review after #3456 is merged~

closes: #3481

### Commit Message / Changelog Entry

N/A feature branch

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/client/cli/query.go [hunk 1]
```diff
@@ -524,3 +524,37 @@ func GetCmdQueryUpgradeError() *cobra.Command {
...
+			upgrade, err := utils.QueryUpgrade(clientCtx, portID, channelID, prove)
+			if err != nil {
+				return err
+			}
+
+			clientCtx = clientCtx.WithHeight(int64(upgrade.ProofHeight.RevisionHeight))
+			return clientCtx.PrintProto(upgrade)
+		},
+	}
+	cmd.Flags().Bool(flags.FlagProve, true, "show proofs for the query results")
```
Comment (colin-axner) [inline modules/core/04-channel/client/cli/query.go:556]: default should probably be false

File: modules/core/04-channel/keeper/grpc_query.go [hunk 2]
```diff
@@ -538,6 +539,29 @@ func (q Keeper) UpgradeErrorReceipt(c context.Context, req *types.QueryUpgradeEr
 	return types.NewQueryUpgradeErrorResponse(receipt, nil, selfHeight), nil
 }
 
+// Upgrade implements the Query/UpgradeSequence gRPC method
+func (q Keeper) Upgrade(c context.Context, req *types.QueryUpgradeRequest) (*types.QueryUpgradeResponse, error) {
```
Comment (colin-axner) [inline modules/core/04-channel/keeper/grpc_query.go:543]:
```suggestion
func (k Keeper) Upgrade(c context.Context, req *types.QueryUpgradeRequest) (*types.QueryUpgradeResponse, error) {
```

**Review by colin-axner — APPROVED** [review]

File: modules/core/04-channel/keeper/grpc_query.go (hunk 2 above)
Comment (chatton) [reply modules/core/04-channel/keeper/grpc_query.go:543]: the linter disagreed with this change, the other methods all use `q` as the receiver

Comment (colin-axner) [reply modules/core/04-channel/keeper/grpc_query.go:518]: I guess we will update it when we merge main

File: modules/core/04-channel/client/cli/query.go [hunk 3]
```diff
@@ -524,3 +524,37 @@ func GetCmdQueryUpgradeError() *cobra.Command {
...
+		),
+		Args: cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clientCtx, err := client.GetClientQueryContext(cmd)
+			if err != nil {
+				return err
+			}
+			portID := args[0]
+			channelID := args[1]
+			prove, _ := cmd.Flags().GetBool(flags.FlagProve)
```
[the remark this answers is not in the document]
Comment (charleenfei) [reply modules/core/04-channel/client/cli/query.go:510]: i asked this before on a similar PR, [@crodriguezvega preferred adding them for consistency ](https://github.com/cosmos/ibc-go/pull/3360#discussion_r1151580785) but i am not strongly opinionated

File: modules/core/04-channel/keeper/grpc_query_test.go [hunk 4]
```diff
@@ -11,6 +11,7 @@ import (
 	"github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
 	"github.com/cosmos/ibc-go/v7/modules/core/exported"
 	ibctesting "github.com/cosmos/ibc-go/v7/testing"
+	"github.com/cosmos/ibc-go/v7/testing/mock"
 )
 
 const doesnotexist = "doesnotexist"
```
Comment (charleenfei) [inline modules/core/04-channel/keeper/grpc_query_test.go:17]: this is unrelated to this PR i guess, but this const is a bit random here if we are just using string values in all the rest of the tests

File: modules/core/04-channel/keeper/grpc_query_test.go [hunk 5]
```diff
@@ -1539,3 +1540,95 @@ func (suite *KeeperTestSuite) TestQueryUpgradeError() {
...
+		{
+			"channel not found",
+			func() {
+				req = &types.QueryUpgradeRequest{
+					PortId:    "test-port-id",
+					ChannelId: "test-channel-id",
+				}
+			},
+			false,
+		},
```
Comment (charleenfei) [inline modules/core/04-channel/keeper/grpc_query_test.go:1591]: should we add one more test for querying an upgrade on a channel that hasn't initiated an upgrade yet?

**Review by charleenfei — APPROVED** [review]
one nit but otherwise nice work!

File: modules/core/04-channel/client/cli/query.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (colin-axner) [reply modules/core/04-channel/client/cli/query.go:510]: If it ain't broke, don't fix it I guess :shrug:

File: modules/core/04-channel/keeper/grpc_query_test.go (hunk 5 above)
Comment (chatton) [reply modules/core/04-channel/keeper/grpc_query_test.go:1591]: this test is actually just directly setting the upgrade in state, the upgrade isn't being set in the ChanUpgradeInit function

Comment (chatton) [reply modules/core/04-channel/keeper/grpc_query_test.go:1591]: actually we can still do this by just deleting the key 👍

**Pull request merged**